schedulefree/radam_schedulefree.py

```diff
@@ -189,11 +197,21 @@ def step(self, closure: Optional[Callable[[], float]] = None) -> Optional[float]
                 # Weight decay calculated at y
                 if decay != 0:
                     torch._foreach_add_(grad, y, alpha=decay)
-
-                # These operations update y in-place,
-                # without computing x explicitly.
-                torch._foreach_lerp_(y, z, weight=ckp1)
-                torch._foreach_add_(y, grad, alpha=adaptive_y_lr)
+                
+                if cautious:
+                    u = torch._foreach_sub(y, z)
+                    torch._foreach_mul_(u, ckp1)
+                    torch._foreach_add_(u, grad, alpha=adaptive_y_lr)
+                    mask = torch._foreach_mul(u, grad)
+                    mask = [(m > 0).to(g.dtype) for m, g in zip(mask, grad)]
+                    torch._foreach_mul_(mask, [m.numel() / (m.sum() + 1) for m in mask])
+                    torch._foreach_mul_(u, mask)
+                    torch._foreach_sub_(y, u)
+                else:
+                    # These operations update y in-place,
+                    # without computing x explicitly.
+                    torch._foreach_lerp_(y, z, weight=ckp1)
+                    torch._foreach_sub_(y, grad, alpha=adaptive_y_lr)
```

Comment:
Hey @nhamanasu, I might be missing something, but is the subtraction correct here (it also appears in the non-foreach and closure versions)? I'm wondering if this is an error that might have been introduced unintentionally.

Reply:
Thank you for the comment!

You're exactly right. In my test branch, I reversed the sign of `adaptive_y_lr` and used `sub` functions, but I somehow forgot to reflect these changes to this c-radam branch. This might have led to completely opposite results. Thank you for catching this critical issue!

Reply:
..sorry, after considering the combination with cautious update, I'm not sure which sign is correct for this part. Let me re-think this block!

Reply:
In the end, I concluded your concern was right. Thank you again for the valuable comments!